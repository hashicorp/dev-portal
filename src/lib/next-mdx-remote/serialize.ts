/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { trace } from '@opentelemetry/api'
import mdx from '@mdx-js/mdx'
import { transform } from '@swc/core'
import remove from 'unist-util-remove'

// types
import { Plugin } from 'unified'
import { MDXRemoteSerializeResult, SerializeOptions } from './types'

/**
 * remark plugin which removes all import and export statements
 */
const removeImportsExportsPlugin: Plugin = () => (tree) =>
	//@ts-expect-error -- remove is a valid method
	remove(tree, ['import', 'export'])

/**
 * Parses and compiles the provided MDX string. Returns a result which can be passed into <MDXRemote /> to be rendered.
 */
export async function serialize(
	/** Raw MDX contents as a string. */
	source: string,
	{ scope = {}, mdxOptions = {}, target = 'es2020' }: SerializeOptions = {}
): Promise<MDXRemoteSerializeResult> {
	// don't modify the original object when adding our own plugin
	// this allows code to reuse the same options object
	const remarkPlugins = [
		...(mdxOptions.remarkPlugins || []),
		removeImportsExportsPlugin,
	]
	mdxOptions = {
		...mdxOptions,
		remarkPlugins,
	}

	const tracer = trace.getTracer('lib/next-mdx-remote')

	const compiledMdx = await tracer.startActiveSpan('mdx', async (span) => {
		try {
			return await mdx(source, { ...mdxOptions, skipExport: true })
		} finally {
			span.end()
		}
	})

	const transformResult = await tracer.startActiveSpan(
		'transform',
		async (span) => {
			try {
				return await transform(compiledMdx, {
					minify: true,
					jsc: {
						parser: {
							syntax: 'ecmascript',
							jsx: true,
						},
						transform: {
							react: {
								pragma: 'mdx',
							},
						},
						target,
					},
				})
			} finally {
				span.end()
			}
		}
	)

	return {
		compiledSource: transformResult.code,
		scope,
	}
}
