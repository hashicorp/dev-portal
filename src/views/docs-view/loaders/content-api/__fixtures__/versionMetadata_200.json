{
	"meta": {
		"status_code": 200,
		"status_text": "OK"
	},
	"result": [
		{
			"product": "waypoint",
			"ref": "kevin/versioned-docs",
			"version": "v0.6.x",
			"created_at": "Fri Sep 24 2021 07:56:45 GMT+0000 (Coordinated Universal Time)",
			"display": "v0.6.2",
			"sha": "e14900057317f9fd63eae31f3414a0ba383faa8e",
			"isLatest": false,
			"releaseStage": "alpha",
			"fullPath": "version-metadata/v0.6.x"
		},
		{
			"product": "waypoint",
			"ref": "kevin/versioned-docs",
			"version": "v0.5.x",
			"created_at": "Thu Sep 23 2021 07:56:45 GMT+0000 (Coordinated Universal Time)",
			"display": "v0.5.2",
			"sha": "214900057317f9fd63eae31f3414a0ba383faa9e",
			"isLatest": true,
			"releaseStage": "stable",
			"fullPath": "version-metadata/v0.5.x"
		},
		{
			"product": "waypoint",
			"ref": "release/0.4.x",
			"version": "v0.4.x",
			"created_at": "Tue Sep 21 2021 21:05:43 GMT+0000 (Coordinated Universal Time)",
			"sha": "c6cd81e10f30d339c113772b5f978d2f96cd1a2e",
			"releaseStage": "stable",
			"fullPath": "version-metadata/v0.4.x"
		},
		{
			"product": "waypoint",
			"ref": "refs/heads/release/0.3.x",
			"version": "v0.3.x",
			"created_at": "Wed Sep 22 2021 17:40:36 GMT+0000 (Coordinated Universal Time)",
			"sha": "61a1d140146321a8d03859ba8b336bd5baa55f67",
			"releaseStage": "stable",
			"fullPath": "version-metadata/v0.3.x"
		}
	]
}
