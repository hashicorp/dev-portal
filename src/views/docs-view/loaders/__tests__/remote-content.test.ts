/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import nock from 'nock'
import document_200 from '../content-api/__fixtures__/document_200.json'
import navData_200 from '../content-api/__fixtures__/navData_200.json'
import document_v4 from '../content-api/__fixtures__/document_v0.4.x.json'
import navData_v4 from '../content-api/__fixtures__/navData_v0.4.x.json'
import versionMetadata_200 from '../content-api/__fixtures__/versionMetadata_200.json'
import RemoteContentLoader, {
	mapVersionList,
	VersionMetadataItem,
} from '../remote-content'

let loader: RemoteContentLoader
let scope: nock.Scope

import * as nextMdxRemote from 'next-mdx-remote-v1/serialize'
const serializeSpy = jest.spyOn(nextMdxRemote, 'serialize')
const mockMdxContentHook = jest.fn()

describe('RemoteContentLoader', () => {
	beforeAll(() => {
		loader = new RemoteContentLoader({
			basePath: 'commands',
			product: 'waypoint',
		})

		nock.disableNetConnect()

		scope = nock(process.env.MKTG_CONTENT_API)
	})

	afterAll(() => {
		nock.enableNetConnect()
	})

	test('generates paths from remote nav data', async () => {
		scope
			.get('/api/content/waypoint/version-metadata')
			.query({ partial: 'true' })
			.reply(200, versionMetadata_200)
		scope
			.get('/api/content/waypoint/nav-data/v0.5.x/commands')
			.reply(200, navData_200)

		const paths = await loader.loadStaticPaths()

		expect(paths).toMatchSnapshot()
	})

	test("allows 'navDataPrefix' to look up nav data", async () => {
		const loader = new RemoteContentLoader({
			basePath: 'plugin/mux',
			navDataPrefix: 'plugin-mux',
			product: 'terraform-plugin-mux',
		})

		scope
			.get('/api/content/terraform-plugin-mux/version-metadata')
			.query({ partial: 'true' })
			.reply(200, {
				meta: {
					status_code: 200,
					status_text: 'OK',
				},
				result: [
					{
						product: 'terraform-plugin-mux',
						ref: 'refs/heads/main',
						version: 'v0.6.x',
						created_at: '2022-05-05T20:45:15.560Z',
						display: 'v0.6.x',
						sha: 'b20cf6618b0bdf4c9eb8895b1d30eca11cc3eae5',
						sk: 'version-metadata/v0.6.x',
						isLatest: true,
						pk: 'terraform-plugin-mux#version-metadata',
					},
				],
			})

		// NOTE: The key assertion is the 'plugin-mux' segment in this URL
		scope
			.get('/api/content/terraform-plugin-mux/nav-data/v0.6.x/plugin-mux')
			.reply(200, {
				meta: {
					status_code: 200,
					status_text: 'OK',
				},
				result: {
					product: 'terraform-plugin-mux',
					githubFile: 'website/docs/plugin-mux-nav-data.json',
					version: 'v0.6.x',
					created_at: '2022-05-05T20:45:15.438Z',
					sha: 'b20cf6618b0bdf4c9eb8895b1d30eca11cc3eae5',
					sk: 'nav-data/v0.6.x/plugin-mux',
					subpath: 'plugin-mux',
					pk: 'terraform-plugin-mux#nav-data/v0.6.x/plugin-mux',
					navData: [
						{
							heading: 'Combining and Translating',
						},
						{
							title: 'Overview',
							path: '',
						},
						{
							title: 'Combining Protocol v5 Providers',
							path: 'combining-protocol-version-5-providers',
						},
						{
							title: 'Combining Protocol v6 Providers',
							path: 'combining-protocol-version-6-providers',
						},
						{
							title: 'Translating Protocol v5 to v6',
							path: 'translating-protocol-version-5-to-6',
						},
						{
							title: 'Translating Protocol v6 to v5',
							path: 'translating-protocol-version-6-to-5',
						},
					],
				},
			})

		const paths = await loader.loadStaticPaths()
		expect(paths).toHaveLength(6)
	})

	test('generates props from remote data', async () => {
		scope
			.get('/api/content/waypoint/version-metadata')
			.query({ partial: 'true' })
			.reply(200, versionMetadata_200)
		scope
			.get('/api/content/waypoint/doc/v0.5.x/commands')
			.reply(200, document_200)
		scope
			.get('/api/content/waypoint/nav-data/v0.5.x/commands')
			.reply(200, navData_200)

		const props = await loader.loadStaticProps({ params: {} })

		expect(props).toMatchInlineSnapshot(
			{
				mdxSource: {
					compiledSource: expect.any(String),
				},

				navData: expect.any(Array),
			},
			`
      Object {
        "currentPath": "",
        "frontMatter": Object {
          "layout": "commands",
          "page_title": "Waypoint Commands (CLI)",
        },
        "githubFileUrl": "https://github.com/hashicorp/waypoint/blob/main/website/content/commands/index.mdx",
        "mdxSource": Object {
          "compiledSource": Any<String>,
          "scope": Object {
            "version": "latest",
          },
        },
        "navData": Any<Array>,
        "versions": Array [
          Object {
            "isLatest": false,
            "label": "v0.6.2 (alpha)",
            "name": "v0.6.x",
            "releaseStage": "alpha",
            "version": "v0.6.x",
          },
          Object {
            "isLatest": true,
            "label": "v0.5.2 (latest)",
            "name": "latest",
            "releaseStage": "stable",
            "version": "v0.5.x",
          },
          Object {
            "isLatest": false,
            "label": "v0.4.x",
            "name": "v0.4.x",
            "releaseStage": "stable",
            "version": "v0.4.x",
          },
          Object {
            "isLatest": false,
            "label": "v0.3.x",
            "name": "v0.3.x",
            "releaseStage": "stable",
            "version": "v0.3.x",
          },
        ],
      }
    `
		)
	})

	test('should return a null github file url for non-latest versions', async () => {
		scope
			.get('/api/content/waypoint/version-metadata')
			.query({ partial: 'true' })
			.reply(200, versionMetadata_200)
		scope
			.get('/api/content/waypoint/doc/v0.4.x/commands')
			.reply(200, document_v4)
		scope
			.get('/api/content/waypoint/nav-data/v0.4.x/commands')
			.reply(200, navData_v4)

		const versionedDocsLoader = new RemoteContentLoader({
			...loader.opts,
			enabledVersionedDocs: true,
		})

		const props = await versionedDocsLoader.loadStaticProps({
			params: {
				page: ['v0.4.x'],
			},
		})

		expect(props.githubFileUrl).toBeNull()
	})

	test('mdxContentHook is called if provided', async () => {
		scope
			.get('/api/content/waypoint/version-metadata')
			.query({ partial: 'true' })
			.reply(200, versionMetadata_200)
		scope
			.get('/api/content/waypoint/doc/v0.4.x/commands')
			.reply(200, document_v4)
		scope
			.get('/api/content/waypoint/nav-data/v0.4.x/commands')
			.reply(200, navData_v4)

		mockMdxContentHook.mockImplementation((mdxContent, scope) => 'Mock impl')

		const versionedDocsLoader = new RemoteContentLoader({
			...loader.opts,
			enabledVersionedDocs: true,
			mdxContentHook: mockMdxContentHook,
			scope: {
				version: 'v0.4.x',
			},
		})

		await versionedDocsLoader.loadStaticProps({
			params: {
				page: ['v0.4.x'],
			},
		})

		expect(mockMdxContentHook).toHaveBeenCalledWith(expect.any(String), {
			version: 'v0.4.x',
		})
		// assert that `serialize` is called with the result of the hook
		expect(serializeSpy).toHaveBeenCalledWith('Mock impl', expect.any(Object))
	})
})

describe('mapVersionList', () => {
	const versionMetadataList = versionMetadata_200.result
	const versionList = mapVersionList(
		versionMetadataList as VersionMetadataItem[]
	)

	test('should label the latest version as "latest"', () => {
		const latestVersion = versionList.find(
			(version) => version.isLatest === true
		)

		expect(latestVersion?.label.endsWith('(latest)')).toBe(true)
	})

	test('should label the alpha version as "alpha"', () => {
		const latestVersion = versionList.find(
			(version) => version.releaseStage === 'alpha'
		)

		expect(latestVersion?.label.endsWith('(alpha)')).toBe(true)
	})

	test('should map a list of version-metadata to a format for <VersionSelect/>', () => {
		expect(versionList).toMatchInlineSnapshot(`
      Array [
        Object {
          "isLatest": false,
          "label": "v0.6.2 (alpha)",
          "name": "v0.6.x",
          "releaseStage": "alpha",
          "version": "v0.6.x",
        },
        Object {
          "isLatest": true,
          "label": "v0.5.2 (latest)",
          "name": "latest",
          "releaseStage": "stable",
          "version": "v0.5.x",
        },
        Object {
          "isLatest": false,
          "label": "v0.4.x",
          "name": "v0.4.x",
          "releaseStage": "stable",
          "version": "v0.4.x",
        },
        Object {
          "isLatest": false,
          "label": "v0.3.x",
          "name": "v0.3.x",
          "releaseStage": "stable",
          "version": "v0.3.x",
        },
      ]
    `)
	})
})
