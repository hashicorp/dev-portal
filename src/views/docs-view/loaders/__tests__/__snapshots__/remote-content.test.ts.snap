// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteContentLoader generates paths from remote nav data 1`] = `
Array [
  Object {
    "params": Object {
      "page": Array [],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "build",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "deploy",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-inspect",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "destroy",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "exec",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "init",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "install",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "login",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "logs",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "project-apply",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "release",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "ui",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "up",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "artifact-build",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "artifact-list-builds",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "artifact-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "artifact-push",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "auth-method-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "auth-method-inspect",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "auth-method-set-oidc",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "auth-method-delete",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "config-get",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "config-set",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "config-source-get",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "config-source-set",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "config-sync",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-clear",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-create",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-delete",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-rename",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-use",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "context-verify",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "deployment-deploy",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "deployment-destroy",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "deployment-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "docs",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "fmt",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "hostname-delete",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "hostname-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "hostname-register",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "plugin",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "project-inspect",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "project-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "release-list",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "runner-agent",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-bootstrap",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-config-set",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-install",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-restore",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-run",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-snapshot",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-uninstall",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "server-upgrade",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "status",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "token-exchange",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "token-invite",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "token-new",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "user-inspect",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "user-invite",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "user-modify",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "user-token",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "version",
      ],
    },
  },
]
`;
